pkg/config/signing/ssh/verify.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	// Now we unmarshal it into the ArmoredSignature struct
+	sig := ArmoredSignature{}
+	if err := ssh.Unmarshal(pemBlock.Bytes, &sig); err != nil {
+		return nil, err
+	}
+
+	if sig.Version != supportedSignatureVersion {
+		return nil, fmt.Errorf("unsupported signature version: %d", sig.Version)
+	}
+	if string(sig.MagicHeader[:]) != magicHeader {
+		return nil, fmt.Errorf("invalid magic header: %s", sig.MagicHeader[:])
+	}
+	if sig.Namespace != signatureFileNamespace {
+		return nil, fmt.Errorf("invalid signature namespace: %s", sig.Namespace)
+	}
+	if _, ok := supportedHashAlgorithms[sig.HashAlgorithm]; !ok {
+		return nil, fmt.Errorf("unsupported hash algorithm: %s", sig.HashAlgorithm)
+	}
+
+	// Now we can unpack the Signature and PublicKey blocks
+	sshSig := ssh.Signature{}
+	if err := ssh.Unmarshal([]byte(sig.Signature), &sshSig); err != nil {
+		return nil, err
+	}
+
+	pk, err := ssh.ParsePublicKey([]byte(sig.PublicKey))
+	if err != nil {
+		return nil, err
+	}
+
+	return &Signature{
+		signature: &sshSig,
+		pk:        pk,
+		hashAlg:   sig.HashAlgorithm,
+	}, nil
+}
+
+func (sv *SignatureVerifier) VerifyJobConfigBytes(
+	jobConfigData []byte,
+	signatureData []byte,
+	allowedSignersData []byte,
+) error {
+	signature, err := decodeArmoredSignature(signatureData)
+
+	if err != nil {
+		return fmt.Errorf("error decoding job config signature: %w", err)
+	}
+
+	h := supportedHashAlgorithms[signature.hashAlg]()
+	h.Write(jobConfigData)
+	jobConfigHash := h.Sum(nil)
+
+	toVerify := MessageWrapper{
+		Namespace:     "file",
+		HashAlgorithm: signature.hashAlg,
+		Hash:          string(jobConfigHash),
+	}
+
+	signedMessage := ssh.Marshal(toVerify)
+	signedMessage = append([]byte(magicHeader), signedMessage...)
+
+	// basic sanity test... does it verify with its own public key
+	err = signature.pk.Verify(signedMessage, signature.signature)
+	if err != nil {
+		log.Printf("job config fails validating against its own pk!")
+	}
+
+	var publicKey ssh.PublicKey
+	var keyComment string
+	var keyOptions []string
+
+	for {
+		publicKey, keyComment, keyOptions, allowedSignersData, err = ssh.ParseAuthorizedKey(allowedSignersData)
+		if err != nil {
+			// by looking at the impl we get an error only when we don't find any more SSH keys
+			return fmt.Errorf("error looking for pk that validates signature: %w", err)
+		}
+
+		// TODO we should probably filter on key type
+		// (makes no sense to try and validate an rsa signature with an ed25519 key)
+
+		log.Printf(
+			"attempting job config signature verification with key %s, comment: %s, options: %s",
+			publicKey.Type(), keyComment, keyOptions,
+		)
+		verificationErr := publicKey.Verify(signedMessage, signature.signature)
+		if verificationErr == nil {
+			log.Print("Job config signature verification succeeded")
+			// Verification succeeded :)
+			return nil
+		}
+		log.Print("Job config signature verification failed")
+	}
```

Comment:
This for loop does not make much sense, does it?
Also, we should return some error message at the end if the verification failed

Reply:
This for loop tries all the public keys that exist in the file to check if any of those verifies the signature.
If we had principal information on the authorized signers file and the signature itself we could check only the matching public key but that's not available in the interfaces (may be available on the concrete implementations) but that would be a change in format that I don't think is supported in openssh